pkg/rhcos/openstack.go

```diff
@@ -25,7 +34,114 @@ func OpenStack(ctx context.Context) (string, error) {
 		return "", err
 	}
 
-	return base.ResolveReference(relOpenStack).String(), nil
+	// Attach uncompressed sha256 checksum to the URL
+	checkSuffix := "?sha256=" + meta.Images.OpenStack.UncompressedSHA256
+	baseURL := base.ResolveReference(relOpenStack).String() + checkSuffix
+
+	// Check that we have generated a valid URL
+	_, err = url.ParseRequestURI(baseURL)
+	if err != nil {
+		return "", err
+	}
+
+	return baseURL, nil
+}
+
+func getCacheDir() (string, error) {
+	// TODO(mfedosin): os.UserCacheDir was added in golang 1.11, but since we still use 1.10
+	// we can't use it directly. Uncomment this when we start using golang 1.11+ for building
+	// the binary.
+	//defaultCacheDir, err := os.UserCacheDir()
+
+	defaultCacheDir := configdir.New("openshift-installer", "images-cache").QueryCacheFolder()
+
+	defaultCacheDir.MkdirAll()
+
+	return defaultCacheDir.Path, nil
+}
+
+func validateSHA256Checksum(filepath, checksum string) error {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+
+	if checksum != fmt.Sprintf("%x", h.Sum(nil)) {
+		logrus.Fatal("Image sha256 checksum is invalid.")
+		defer os.Remove(filepath)
+		return errors.Errorf("image sha256 checksum is invalid")
+	}
+
+	return nil
+}
+
+// DownloadOpenStackImageFile obtains a file from a given URL, puts it in the cache
+// and returns the local file path
+func DownloadOpenStackImageFile(baseURL string) (string, error) {
+	logrus.Infof("Obtaining RHCOS image file from '%v'", baseURL)
+
+	// Convert the given URL into a file name using md5 algorithm
+	imageFileName := fmt.Sprintf("%x.img", md5.Sum([]byte(baseURL)))
+
+	// If the file has already been cached, return its path
+	cacheDir, err := getCacheDir()
+	if err != nil {
+		return "", err
+	}
+	filePath := filepath.Join(cacheDir, imageFileName)
+
+	_, err = os.Stat(filePath)
+	if !os.IsNotExist(err) {
```

Comment:
Er, this means we return successfully if we get e.g. `EPERM` or `EIO`.  This type of issue is why I'm a Rust fan, see https://github.com/cgwalters/openat-ext/blob/143c32354eda317133ea1cefb4b1aeee9766c91c/src/lib.rs#L32
Anyways, for Go code see e.g. https://github.com/openshift/machine-config-operator/blob/5f1f4b9487df239ea850ddd4ba7499926409c770/pkg/daemon/node.go#L27